src/runtime/recipe/walker.ts

```diff
@@ -121,41 +140,44 @@ export abstract class Walker<T extends Cloneable> {
     return walker.descendants;
   }
 
-  _runUpdateList(start: T, updateList) {
-    const updated = [];
+  visit<Ctx extends object[]>(visitor: (obj: T, ...ctx: Ctx) => Continuation<T, Ctx>, ...context: Ctx): void {
+    const continuation: Continuation<T, Ctx> = visitor.bind(this)(this.currentResult.result, ...context);
+    if (!this.isEmptyResult(continuation)) {
+      this.updateList.push({continuation, context});
+    }
+  }
+
+  private runUpdateList(start: T, updateList: Update<T, object[]>[]) {
+    const updated: {result: T, score: number}[] = [];
     if (updateList.length) {
       switch (this.tactic) {
         case WalkerTactic.Permuted: {
-          let permutations = [[]];
+          let permutations: SingleUpdate<T, object[]>[][] = [[]];
           updateList.forEach(({continuation, context}) => {
-            const newResults = [];
+            const newResults: SingleUpdate<T, object[]>[][] = [];
             if (typeof continuation === 'function') {
               continuation = [continuation];
             }
             continuation.forEach(f => {
               permutations.forEach(p => {
                 const newP = p.slice();
-                newP.push({f, context});
+                newP.push({continuation: f, context});
                 newResults.push(newP);
               });
             });
             permutations = newResults;
           });
 
           for (let permutation of permutations) {
-            const cloneMap = new Map();
+            const cloneMap = new Map<object, object>();
```

Comment:
I've been told to prefer typing this like so:

```
const cloneMap: Map<object, object> = new Map();
```

Is one style better than the other?

Reply:
Discussed f2f, we think it's fine as is, or even preferred given it's shorter.